File: 2428-equal-row-and-column-pairs/2428-equal-row-and-column-pairs.py
class Solution:
    def equalPairs(self, grid: List[List[int]]) -> int:
        length = len(grid)
        map1 = { }
        map2 = { }

        for i in range(length):
            s = []
            for j in range(length):
                s.append(str(grid[i][j]))
                s.append(":")
            str_row = "".join(s)
            map1[str_row] = map1.get(str_row, 0) + 1

        for i in range(length):
            s = []
            for j in range(length):
                s.append(str(grid[j][i]))
                s.append(":")
            str_col = "".join(s)
            map2[str_col] = map2.get(str_col, 0)+1
        
        ans = 0
        for s in map1.keys():
            if s in map2:
                ans += map1[s] * map2[s]

        return ans
        

File: 2428-equal-row-and-column-pairs/2428-equal-row-and-column-pairs.java
class Solution {
    public int equalPairs(int[][] grid) {
        int n = grid.length;
        HashMap<String, Integer> map1 = new HashMap<>();
        HashMap<String, Integer> map2 = new HashMap<>();

        for(int i = 0; i<n; i++){
            StringBuilder sb = new StringBuilder();
            for(int j = 0; j<n; j++){
                sb.append(grid[i][j]);
                sb.append(":");
            }
            String str = sb.toString();
            map1.put(str, map1.getOrDefault(str, 0)+1);
        }

        for(int i = 0; i<n; i++){
            StringBuilder sb = new StringBuilder();
            for(int j = 0; j<n; j++){
                sb.append(grid[j][i]);
                sb.append(":");
            }
            String str = sb.toString();
            map2.put(str, map2.getOrDefault(str, 0)+1);
        }

        int ans = 0;
        for(String st : map1.keySet()){
            if(map2.containsKey(st)){
                ans += map1.get(st) * map2.get(st);
            }            
        }
        return ans;
    }
}
